File: src/Systems/C++/proxy.cpp
#include <iostream>
#include <memory>
#include <string>
#include <unordered_map>

class DocumentStore {
public:
    virtual ~DocumentStore() = default;
    virtual std::string get(int id) = 0;
};

class RemoteDocumentStore final : public DocumentStore {
public:
    std::string get(int id) override {
        ++fetch_count_;
        return "doc(" + std::to_string(id) + ")";
    }

    int fetch_count() const { return fetch_count_; }

private:
    int fetch_count_ = 0;
};

class DocumentStoreProxy final : public DocumentStore {
public:
    std::string get(int id) override {
        if (const auto it = cache_.find(id); it != cache_.end()) {
            return it->second;
        }
        if (!backend_) {
            backend_ = std::make_unique<RemoteDocumentStore>();
        }
        auto value = backend_->get(id);
        cache_.emplace(id, value);
        return value;
    }

    int backend_count() const { return backend_ ? 1 : 0; }
    int fetch_count() const { return backend_ ? backend_->fetch_count() : 0; }

private:
    std::unique_ptr<RemoteDocumentStore> backend_;
    std::unordered_map<int, std::string> cache_;
};

int main() {
    DocumentStoreProxy store;
    const auto first = store.get(42);
    const auto second = store.get(42);
    std::cout << "backend=" << store.backend_count()
              << ";fetches=" << store.fetch_count()
              << ";first=" << first
              << ";second=" << second << '\n';
}

File: src/Systems/Rust/proxy.rs
use std::collections::HashMap;

trait DocumentStore {
    fn get(&mut self, id: u32) -> String;
}

struct RemoteDocumentStore {
    fetch_count: u32,
}

impl RemoteDocumentStore {
    fn new() -> Self {
        Self { fetch_count: 0 }
    }
}

impl DocumentStore for RemoteDocumentStore {
    fn get(&mut self, id: u32) -> String {
        self.fetch_count += 1;
        format!("doc({id})")
    }
}

struct DocumentStoreProxy {
    backend: Option<RemoteDocumentStore>,
    cache: HashMap<u32, String>,
}

impl DocumentStoreProxy {
    fn new() -> Self {
        Self { backend: None, cache: HashMap::new() }
    }

    fn backend_count(&self) -> u32 {
        u32::from(self.backend.is_some())
    }

    fn fetch_count(&self) -> u32 {
        self.backend.as_ref().map_or(0, |backend| backend.fetch_count)
    }
}

impl DocumentStore for DocumentStoreProxy {
    fn get(&mut self, id: u32) -> String {
        if let Some(value) = self.cache.get(&id) {
            return value.clone();
        }
        if self.backend.is_none() {
            self.backend = Some(RemoteDocumentStore::new());
        }
        let value = self.backend.as_mut().expect("backend created").get(id);
        self.cache.insert(id, value.clone());
        value
    }
}

fn main() {
    let mut store = DocumentStoreProxy::new();
    let first = store.get(42);
    let second = store.get(42);
    println!(
        "backend={};fetches={};first={};second={}",
        store.backend_count(), store.fetch_count(), first, second
    );
}
